File: engine/src/domain/job.cpp
#include "flowforge/domain/job.hpp"

namespace flowforge::domain {

std::string_view to_string(JobStatus status) noexcept {
  switch (status) {
    case JobStatus::Pending:
      return "pending";
    case JobStatus::Queued:
      return "queued";
    case JobStatus::Running:
      return "running";
    case JobStatus::Succeeded:
      return "succeeded";
    case JobStatus::Failed:
      return "failed";
    case JobStatus::Retrying:
      return "retrying";
    case JobStatus::Cancelled:
      return "cancelled";
    case JobStatus::DeadLetter:
      return "dead_letter";
  }
  return "unknown";
}

bool is_terminal(JobStatus status) noexcept {
  switch (status) {
    case JobStatus::Succeeded:
    case JobStatus::Cancelled:
    case JobStatus::DeadLetter:
      return true;
    default:
      return false;
  }
}

std::optional<JobStatus> job_status_from_string(std::string_view value) noexcept {
  if (value == "pending")
    return JobStatus::Pending;
  if (value == "queued")
    return JobStatus::Queued;
  if (value == "running")
    return JobStatus::Running;
  if (value == "succeeded")
    return JobStatus::Succeeded;
  if (value == "failed")
    return JobStatus::Failed;
  if (value == "retrying")
    return JobStatus::Retrying;
  if (value == "cancelled")
    return JobStatus::Cancelled;
  if (value == "dead_letter")
    return JobStatus::DeadLetter;
  return std::nullopt;
}

Job Job::restore(infra::JobId id, std::string queue_name, std::string payload, RetryPolicy retry_policy,
                 int priority, JobStatus status, std::uint32_t attempt_count,
                 std::optional<std::string> last_error, infra::TimePoint created_at,
                 infra::TimePoint updated_at, std::string job_type,
                 std::optional<infra::WorkloadId> workload_id) {
  Job job(std::move(id), std::move(queue_name), std::move(payload), retry_policy, created_at, priority,
          std::move(job_type), std::move(workload_id));
  job.status_ = status;
  job.attempt_count_ = attempt_count;
  job.last_error_ = std::move(last_error);
  job.updated_at_ = updated_at;
  return job;
}

}  // namespace flowforge::domain

File: packages/shared/src/job.ts
/**
 * Mirrors flowforge::domain::JobStatus (engine/include/flowforge/domain/job.hpp)
 * as serialized by apps/server/src/json/job_json.cpp.
 */
export type JobStatus =
  | "pending"
  | "queued"
  | "running"
  | "succeeded"
  | "failed"
  | "retrying"
  | "cancelled"
  | "dead_letter";

export interface Job {
  id: string;
  queue_name: string;
  /** "" means not schedulable yet -- see docs/architecture/execution-model.md. */
  job_type: string;
  payload: unknown;
  priority: number;
  status: JobStatus;
  attempt_count: number;
  max_attempts: number;
  last_error: string | null;
  created_at: string;
  updated_at: string;
}

export interface RetryPolicyInput {
  max_attempts?: number;
  initial_backoff_ms?: number;
  max_backoff_ms?: number;
  backoff_multiplier?: number;
}

export interface CreateJobInput {
  queue_name: string;
  payload: unknown;
  /** Optional. Non-empty + registered (e.g. "echo") submits the job to the Scheduler on creation. */
  job_type?: string;
  priority?: number;
  retry_policy?: RetryPolicyInput;
}

/** Additive response field from POST /api/v1/jobs (Phase 2B-2) reporting
 *  whether the Scheduler accepted the job -- see
 *  docs/architecture/execution-model.md, "HTTP integration". */
export interface SchedulingOutcome {
  scheduled: boolean;
  reason?: string;
}

export interface CreateJobResponse extends Job {
  scheduling: SchedulingOutcome;
}

export interface ListJobsResponse {
  jobs: Job[];
}

/**
 * Mirrors flowforge::domain::ExecutionOutcome (engine/include/flowforge/domain/execution.hpp).
 */
export type ExecutionOutcome = "running" | "succeeded" | "failed" | "timed_out" | "cancelled";

/**
 * One real execution attempt (Phase 2B-3) -- mirrors flowforge::domain::Execution,
 * as serialized by apps/server/src/json/job_json.cpp's `to_json(const domain::Execution&)`.
 * Fetched via `GET /api/v1/jobs/{id}/attempts`.
 */
export interface Attempt {
  id: string;
  job_id: string;
  worker_id: string | null;
  attempt_number: number;
  outcome: ExecutionOutcome;
  started_at: string;
  finished_at: string | null;
  error_message: string | null;
}

export interface ListAttemptsResponse {
  attempts: Attempt[];
}
